releases-implementation/README.md

```diff
@@ -0,0 +1,415 @@
+    version: 5.1.0
+    dependsOn:
+    - prometheus-operator-crd
+  - name: vertical-pod-autoscaler-crd
+    version: 3.0.0
+  - name: etcd-k8s-res-count-exporter
+    version: 1.10.0
+    dependsOn:
+    - kyverno
+  - name: observability-bundle
+    version: 1.3.4
+    dependsOn:
+    - coredns
+  - name: k8s-dns-node-cache
+    version: 2.6.2
+    dependsOn:
+    - kyverno
+  - name: security-bundle
+    version: 1.6.5
+    catalog: giantswarm
+    dependsOn:
+    - prometheus-operator-crd
+  - name: teleport-kube-agent
+    version: 0.9.0
+  components:
+  - name: cluster-aws
+    version: 1.0.0
+  - name: flatcar
+    version: 3815.2.2
+  - name: flatcar-variant
+    version: 3.0.0
+  - name: kubernetes
+    version: 1.25.16
+  date: "2024-05-18T12:57:50Z"
+  state: active
+```
+
+New releases MUST BE added via pull requests in a similar way like for vintage releases. After a pull request is merged, CircleCI job pushes the newly added Releases to the releases catalog and to the provider-specific app collection.
+
+#### 4.1.2. Deploying a cluster
+
+Today workload clusters in our Cluster API-based KaaS product are deployed with cluster-$provider apps. That remains the same, with only one minor difference, which is how we specify the version, where instead of specifying App’s `.spec.version` property we MUST specify `release.giantswarm.io/version` label and leave `.spec.version` empty.
+
+So the whole manifest for the cluster-$provider app and its config would look like this:
+
+```yaml
+---
+apiVersion: v1
+data:
+  values: |
+    global:
+      metadata:
+        name: mycluster
+        organization: mycompany
+        description: Production cluster
+kind: ConfigMap
+metadata:
+  creationTimestamp: null
+  labels:
+    giantswarm.io/cluster: mycluster
+  name: mycluster-userconfig
+  namespace: org-mycompany
+---
+apiVersion: application.giantswarm.io/v1alpha1
+kind: App
+metadata:
+  labels:
+    app-operator.giantswarm.io/version: 0.0.0
+    release.giantswarm.io/version: 25.0.0
+  name: mycluster
+  namespace: org-mycompany
+spec:
+  catalog: cluster
+  kubeConfig:
+    inCluster: true
+  name: cluster-aws
+  namespace: org-mycompany
+  userConfig:
+    configMap:
+      name: mycluster-userconfig
+      namespace: org-mycompany
+  version: ""
+```
+
+After cluster-$provider app is deployed, few things will happen:
+- an App mutating webhook reads the release version, looks up the corresponding release, reads cluster-$provider app version from it and finally it sets App’s `.spec.version` property,
+- after cluster-$provider app is applied successfully, in order to render all templates and set the app version, catalog and depends on properties, Helm will do the following:
+	- it will lookup the App resource (via Helm `lookup` function) and read release version,
+	- then it will lookup the Release resource and read app version, catalog and depends on properties from there, and
+	- finally it will render all templates and apply them.
```

Comment:
(Thinking out loud...) I wonder if it might be better / easier to instead have a ConfigMap deployed with each Release that acts like a values file that can be used as an overlay for each cluster-$provider App instead of needing to do multiple helm lookups.

Reply:
+1 not sure if this is upstream helm functionality and if that is supported by all helm implementations (e.g. within Flux, Argo,...)

Reply:
> I wonder if it might be better / easier to instead have a ConfigMap deployed with each Release that acts like a values file that can be used as an overlay for each cluster-$provider App instead of needing to do multiple helm lookups.

Better, not sure, maybe yes, maybe not.

Easier? I don't think so, at least not now, because we already have process, workflows, tooling for the releases repo and Release CRs, and one of the ideas here is to make use of at least some of that, because otherwise we are taking more than few steps back as we're dropping the tooling and processes that have been working for many years.

Another downside of the ConfigMap idea and using it as an overlay is that it means that k8s/OS/app versions are then Helm values again, and we don't want that, because it makes k8s/OS/app versions part of the workload cluster creation API like it's just a config.

> not sure if this is upstream helm functionality

Helm `lookup` function (which is basically a k8s client) is a regular upstream Helm function https://helm.sh/docs/chart_template_guide/functions_and_pipelines/#using-the-lookup-function, we already use it in some places in cluster-$provider apps.